functorch/_src/aot_autograd.py

```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
         )
         node.register_hook(get_posthook(special_stack))
 
+# This class encapsulates all aliasing + mutation info we need about the forward graph
+# See a more detailed overview of the edge case handling at
+# https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
+@dataclass(frozen=True)
+class ViewAndMutationMeta:
+    # length: # forward inputs
+    # For every input, tells us whether the input:
+    # (a) is not mutated
+    # (b) only metadata is mutated
+    # (c) data (and maybe metadta) is mutated
```

Comment:
typo

Reply:
It's not clear to me why data mutation should subsume metadata mutation. They're pretty different beasts and it seems reasonable to track them separately. I'll be keeping an eye on this.

Reply:
In particular, with this scheme you are obligated to update metadata whenever there is a data mutation, but that might be wasted work if the metadata didn't change

Reply:
(very long delay - I have a large local refactor of AOT Autograd that does a bunch of cleanup, going to finally respond to PR feedback now)

This is true - the way this is handled on master now is in the MutationType.data case, we do a check at runtime to see if the metadata is different. I don't think this is too expensive (we're comparing sizes + strides + storage_offset), but agreed that precomputing a 4th enum would be cheaper.